modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -817,5 +818,880 @@ int HashTSDFVolumeCPU::getVisibleBlocks(int currFrameId, int frameThreshold) con
+    for (int i = 0; i < depth_rows; i++)
+        *y.ptr<float>(0, i) = (i - cxy[1]) / fxy[1];
+
+    cv::String errorStr;
+    cv::String name = "preCalculationPixNorm";
+    ocl::ProgramSource source = ocl::rgbd::hash_tsdf_oclsrc;
+    cv::String options = "-cl-mad-enable";
+    ocl::Kernel kk;
+    kk.create(name.c_str(), source, options, &errorStr);
+
+
+    if (kk.empty())
+        throw std::runtime_error("Failed to create kernel: " + errorStr);
+
+    AccessFlag af = ACCESS_READ;
+    UMat pixNorm = _pixNorm.getUMat(af);
+    UMat xx = x.getUMat(af);
+    UMat yy = y.getUMat(af);
+
+    kk.args(ocl::KernelArg::PtrReadWrite(pixNorm),
+        ocl::KernelArg::PtrReadOnly(xx),
+        ocl::KernelArg::PtrReadOnly(yy),
+        depth_cols);
+
+    size_t globalSize[2];
+    globalSize[0] = depth_rows;
+    globalSize[1] = depth_cols;
+
+    if (!kk.run(2, globalSize, NULL, true))
+        throw std::runtime_error("Failed to run kernel");
+
+    return pixNorm;
+}
+
+void HashTSDFVolumeGPU::integrateAllVolumeUnitsGPU(InputArray _depth, float depthFactor, const Intr& intrinsics)
+{
+    CV_TRACE_FUNCTION();
+    CV_Assert(!_depth.empty());
+
+    UMat depth = _depth.getUMat();
+
+    String errorStr;
+    String name = "integrateAllVolumeUnits";
+    ocl::ProgramSource source = ocl::rgbd::hash_tsdf_oclsrc;
+    String options = "-cl-mad-enable";
+    ocl::Kernel k;
+    k.create(name.c_str(), source, options, &errorStr);
+
+    if (k.empty())
+        throw std::runtime_error("Failed to create kernel: " + errorStr);
+
+    float dfac = 1.f / depthFactor;
+    Vec4i volResGpu(volumeUnitResolution, volumeUnitResolution, volumeUnitResolution);
+    Vec2f fxy(intrinsics.fx, intrinsics.fy), cxy(intrinsics.cx, intrinsics.cy);
+
+    int totalVolUnitsSize = (int) indexes.indexesGPU.size();
+    Mat totalVolUnits = cv::Mat(indexes.indexesGPU, true);
+
+    Mat _tmp;
+    volUnitsData.copyTo(_tmp);
+    UMat U_volUnitsData = _tmp.getUMat(ACCESS_RW);
+    _tmp.release();
+
+    indexes.volumes.copyTo(_tmp);
+    UMat U_hashtable = _tmp.getUMat(ACCESS_RW);
+
+
+    k.args(ocl::KernelArg::ReadOnly(depth),
+        ocl::KernelArg::PtrReadWrite(U_hashtable),
+        (int)indexes.list_size,
+        (int)indexes.bufferNums,
+        (int)indexes.hash_divisor,
+        ocl::KernelArg::ReadWrite(totalVolUnits.getUMat(ACCESS_RW)),
+        ocl::KernelArg::ReadWrite(U_volUnitsData),
+        ocl::KernelArg::PtrReadOnly(pixNorms),
+        ocl::KernelArg::ReadOnly(allVol2cam.getUMat(ACCESS_READ)),
+        lastVolIndex,
+        voxelSize,
+        volResGpu.val,
+        volStrides.val,
+        fxy.val,
+        cxy.val,
+        dfac,
+        truncDist,
+        int(maxWeight)
+    );
+
+    int resol = volumeUnitResolution;
+    size_t globalSize[3];
+    globalSize[0] = (size_t)resol; // volumeUnitResolution
+    globalSize[1] = (size_t)resol; // volumeUnitResolution
+    globalSize[2] = (size_t)totalVolUnitsSize; // num of voxels
```

Comment:
When processing volume units in parallel, we may have multiple parallel queries to far memory locations.
To avoid this, a loop over volume units should be done inside a kernel.
Please check both options: which one is faster.